components/motor/gpio/motor_encoder.go

```diff
@@ -458,7 +461,16 @@ func (m *EncodedMotor) GoTo(ctx context.Context, rpm, targetPosition float64, ex
 
 // ResetZeroPosition sets the current position (+/- offset) to be the new zero (home) position.
 func (m *EncodedMotor) ResetZeroPosition(ctx context.Context, offset float64, extra map[string]interface{}) error {
-	return m.encoder.ResetPosition(ctx, extra)
+	if err := m.Stop(ctx, extra); err != nil {
+		return err
+	}
+	if err := m.encoder.ResetPosition(ctx, extra); err != nil {
+		return err
+	}
+	m.stateMu.Lock()
+	defer m.stateMu.Unlock()
+	m.offsetInTicks = -1 * offset * m.ticksPerRotation
+	return nil
```

Comment:
I do not like how complicated this has made locked and unlocking mutexes. Let's try to think of a better way to do this using the ResetZeroPosition function and perhaps a different struct handling the motor's posito or mutex that protects the position only, so we're not overloading the same mutex for so many things and running into races.

Reply:
yeah valid. I think I just didn't do this in a great way tho, and tbh not sure how there wasn't an issue before because `position` was being locked sometimes and not locked other times so now I've moved the call to `position` in `rpmMonitor` which creates some more consistency and only one new lock/unlock. let me know if that's better but if not I'll try to think of other ways to go about this